alacritty/src/renderer/text/gles2.rs

```diff
@@ -0,0 +1,494 @@
+use crate::renderer::shader::{ShaderProgram, ShaderVersion};
+use crate::renderer::{cstr, Error};
+
+use super::atlas::{Atlas, ATLAS_SIZE};
+use super::{
+    Glyph, GlyphCache, LoadGlyph, LoaderApi, TextRenderApi, TextRenderBatch, TextRenderer,
+};
+
+// Shader source.
+static TEXT_SHADER_F: &str = include_str!("../../../res/gles2/text.f.glsl");
+static TEXT_SHADER_V: &str = include_str!("../../../res/gles2/text.v.glsl");
+
+#[derive(Debug)]
+pub struct Gles2Renderer {
+    program: TextShaderProgram,
+    vao: GLuint,
+    vbo: GLuint,
+    ebo: GLuint,
+    atlas: Vec<Atlas>,
+    batch: Batch,
+    current_atlas: usize,
+    active_tex: GLuint,
+}
+
+impl Gles2Renderer {
+    pub fn new() -> Result<Self, Error> {
+        info!("Using OpenGL ES 2.0 renderer");
+
+        let program = TextShaderProgram::new(ShaderVersion::Gles2)?;
+        let mut vao: GLuint = 0;
+        let mut vbo: GLuint = 0;
+        let mut ebo: GLuint = 0;
+
+        let mut indices = Vec::with_capacity(BATCH_MAX / 4 * 6);
+        for index in 0..(BATCH_MAX / 4) as u16 {
+            let index = index * 4;
+            indices.push(index);
+            indices.push(index + 1);
+            indices.push(index + 3);
+
+            indices.push(index + 1);
+            indices.push(index + 2);
+            indices.push(index + 3);
+        }
+
+        unsafe {
+            gl::Enable(gl::BLEND);
+
+            gl::DepthMask(gl::FALSE);
+
+            gl::GenVertexArrays(1, &mut vao);
+            gl::GenBuffers(1, &mut ebo);
+            gl::GenBuffers(1, &mut vbo);
+            gl::BindVertexArray(vao);
+
+            // Elements buffer.
+            gl::BindBuffer(gl::ELEMENT_ARRAY_BUFFER, ebo);
+            gl::BufferData(
+                gl::ELEMENT_ARRAY_BUFFER,
+                (indices.capacity() * size_of::<u16>()) as isize,
+                indices.as_ptr() as *const _,
+                gl::STATIC_DRAW,
+            );
+
+            // Vertex buffer.
+            gl::BindBuffer(gl::ARRAY_BUFFER, vbo);
+            gl::BufferData(
+                gl::ARRAY_BUFFER,
+                (BATCH_MAX * size_of::<TextVertex>()) as isize,
+                ptr::null(),
+                gl::STREAM_DRAW,
+            );
+
+            let mut index = 0;
+            let mut size = 0;
+
+            macro_rules! add_attr {
+                ($count:expr, $gl_type:expr, $type:ty) => {
+                    gl::VertexAttribPointer(
+                        index,
+                        $count,
+                        $gl_type,
+                        gl::FALSE,
+                        size_of::<TextVertex>() as i32,
+                        size as *const _,
+                    );
+                    gl::EnableVertexAttribArray(index);
+
+                    #[allow(unused_assignments)]
+                    {
+                        size += $count * size_of::<$type>();
+                        index += 1;
+                    }
+                };
+            }
```

Comment:
Was considering to suggest abstracting this macro but it's so short that I don't think it's worth it considering one has the divisor while the other doesn't.